pkg/spdk/engine.go

```diff
@@ -1128,40 +1131,118 @@ func (e *Engine) ReplicaShallowCopy(dstReplicaName, dstReplicaAddress string) (e
+		return err
+	}
+
+	e.log.Infof("Engine copied all snapshots from rebuilding src replica %s to rebuilding dst replica %s", srcReplicaName, dstReplicaName)
+
+	return nil
+}
+
+func (e *Engine) ReplicaAddFinish(spdkClient *spdkclient.Client, srcReplicaName, dstReplicaName, dstReplicaAddress string) (err error) {
+	updateRequired := false
+
+	e.Lock()
+	defer func() {
+		e.Unlock()
+
+		if updateRequired {
+			e.UpdateCh <- nil
+		}
+	}()
+
+	// Syncing with the SPDK TGT server only when the engine is running.
+	if e.State != types.InstanceStateRunning {
+		return fmt.Errorf("invalid state %v for engine %s replica %s add finish", e.State, e.Name, srcReplicaName)
+	}
+	if _, exists := e.ReplicaAddressMap[srcReplicaName]; !exists {
+		return fmt.Errorf("replica %s does not exist in engine %s", srcReplicaName, e.Name)
+	}
+	if e.ReplicaModeMap[dstReplicaName] != types.ModeWO {
+		return fmt.Errorf("invalid mode %s for engine %s replica %s add finish", e.ReplicaModeMap[dstReplicaName], e.Name, dstReplicaName)
+	}
+
+	// TODO: Can we share the clients in the whole server?
+	srcReplicaAddress := e.ReplicaAddressMap[srcReplicaName]
+	srcReplicaServiceCli, err := GetServiceClient(srcReplicaAddress)
+	if err != nil {
+		return err
+	}
+	dstReplicaServiceCli, err := GetServiceClient(dstReplicaAddress)
+	if err != nil {
+		return err
+	}
+
+	var engineErr error
+	defer func() {
+		if engineErr != nil {
+			if e.State != types.InstanceStateError {
+				e.State = types.InstanceStateError
+				updateRequired = true
 			}
-			if err = srcReplicaServiceCli.ReplicaRebuildingSrcDetach(srcReplicaName, dstReplicaName); err != nil {
-				return err
+			e.ErrorMsg = err.Error()
+			if err == nil {
+				err = engineErr
 			}
-			if err = dstReplicaServiceCli.ReplicaRebuildingDstSnapshotRevert(dstReplicaName, snapshotName); err != nil {
-				return err
+		} else {
+			if e.State != types.InstanceStateError {
+				e.ErrorMsg = ""
+			}
+		}
+
+		// Blindly mark the rebuilding replica as mode ERR now.
+		if err != nil {
+			// Blindly send rebuilding finish for src replica.
+			if srcReplicaErr := srcReplicaServiceCli.ReplicaRebuildingSrcFinish(srcReplicaName, dstReplicaName); srcReplicaErr != nil {
+				e.log.WithError(srcReplicaErr).Errorf("Failed to finish rebuilding for src replica %s with address %s after rebuilding failure, will do nothing", srcReplicaName, srcReplicaAddress)
+			}
+			if e.ReplicaModeMap[dstReplicaName] != types.ModeERR {
+				e.ReplicaModeMap[dstReplicaName] = types.ModeERR
+				e.log.WithError(err).Errorf("Failed to finish rebuilding dst replica %s with address %s from src replica %s with address %s, will mark the rebuilding replica mode as ERR", dstReplicaName, dstReplicaAddress, srcReplicaName, srcReplicaAddress)
+				updateRequired = true
 			}
-			break
 		}
+	}()
+
+	// Pause the IO again by suspending the NVMe initiator
+	if e.Frontend == types.FrontendSPDKTCPBlockdev {
+		if err = e.initiator.Suspend(false, true); err != nil {
```

Comment:
Does it need to flush IOs in the suspend?

Reply:
Should we set `egineErr=err` if `Suspend()` failed?

Reply:
When the suspend fails in the middle, I am not sure if the engine still works. Hence I prefer to consider the engine as ERR first.

Reply:
> Does it need to flush IOs in the suspend?

The input `noflush` is set to `false` when calling `e.initiator.Suspend(false, true)`. I would assume the initiator can handle it automatically.

Reply:
I see. There is no need to flush the IO here